src/src/PluginStructs/P014_data_struct.cpp

```diff
@@ -175,148 +249,279 @@ uint8_t P014_data_struct::checkCRC(uint16_t data, uint8_t check)
-  if ((datatype == SI7021_MEASURE_HUM) || (datatype == SI7021_MEASURE_HUM_HM)) {
-    // Convert value to Himidity percent
-    // pm-cz: it is possible to enable decimal places for humidity as well by multiplying the value in formula by 100
-    data = ((1250 * (long)raw) >> 16) - 60;
+  return true;
+}
 
-    // Datasheet says doing this check
-    if (data > 1000) { data = 1000; }
+bool P014_data_struct::disablePowerForADC(uint8_t i2caddr){
+  //set VOUT
+  //Get the current register value
 
-    if (data < 0) { data = 0; }
+  //we know the type of circuit used based on the i2c address 
 
-    // pm-cz: Let us make sure we have enough precision due to ADC bits
-    if (resolution == SI7021_RESOLUTION_12T_08RH) {
-      data  = (data + 5) / 10;
-      data *= 10;
+
+  bool ok=false;
+  uint8_t reg = I2C_read8_reg(i2caddr,SI7013_READ_REG2, &ok);
+
+  if (!ok) {
+    addLog(LOG_LEVEL_ERROR, F("SI7013: Could not read REG2!"));
+    return false;
+  }
+    
+  
+  if (i2caddr == SI7013_I2C_ADDRESS_AD0_1){
+    ok = I2C_write8_reg(i2caddr,SI7013_WRITE_REG2,reg | (1+2+4+64) );//set last three bits to 1 (VIN bufered, Vref=VDD, VOUT=VDD) and No-Hold for bit 6
+  }else{
+    ok = I2C_write8_reg(i2caddr,SI7013_WRITE_REG2, (reg & B11111000) | (2+4+64) );//set last three bits (VIN bufered, Vref=VDD, VOUT=GND) and No-Hold for bit 6
+  }
+  if (!ok){
+    addLog(LOG_LEVEL_ERROR, F("SI7013: Could not write REG2!"));
+    return false;
+  }
+
+  return true;
+}
+
+bool P014_data_struct::requestADC(uint8_t i2caddr)
+{
+    //start ADC conversion
+    if(!I2C_write8(i2caddr, SI7013_READ_ADC)){
+      addLog(LOG_LEVEL_ERROR,F("SI7013: could not start ADC conversion"));
+      return false;
     }
+    //delay(10); //waiting for conversion to be done: in the specs is mentioned 7ms in normal mode
+  return true;
+}
 
-    // save value
-    si7021_humidity = (uint16_t)data;
-
-    // Temperature
-  } else if ((datatype == SI7021_MEASURE_TEMP) || (datatype == SI7021_MEASURE_TEMP_HM) || (datatype == SI7021_MEASURE_TEMP_HUM)) {
-    // Convert value to Temperature (*100)
-    // for 23.45C value will be 2345
-    data =  ((17572 * (long)raw) >> 16) - 4685;
-
-    /*
-       // pm-cz: We should probably check for precision here as well
-       if (resolution != SI7021_RESOLUTION_14T_12RH) {
-       if (data > 0) {
-        data = (data + 5) / 10;
-       } else {
-        data = (data - 5) / 10;
-       }
-       data *= 10;
-       }
-     */
 
-    // save value
-    si7021_temperature = (int16_t)data;
+bool P014_data_struct::readADC(uint8_t i2caddr, uint8_t filter_power)
+{
+  if ( Wire.requestFrom(i2caddr, 2u) < 2 ) {
+    return false;
   }
 
-  return 0;
+  // Comes back in two bytes, data(MSB) / data(LSB) with no Checksum
+  uint16_t raw  = ((uint16_t) Wire.read()) << 8;
+  raw |= Wire.read();
+
+  //Calculate Moving average where 2^filter_power is the moving window of points
+  //MA*[i]= MA*[i-1] +X[i] - MA*[i-1]/N
+  adc = adc + raw - (adc>>filter_power);
```

Comment:
I don't understand this moving average filter.

It looks a bit like an IIR filter, but this keeps increasing.

Reply:
not sure I understand the question.
there is a + raw that adds the current raw value then a - that removes the average of the previous value.
N  = 2^filter_power 
This makes the math much simpler and faster.

Reply:
I do understand the bitshift is a much faster divide operator.
But I don't see how this does an average or a moving average.

Let's assume the value is always 16 (or any power of 2 value).

`adc = 16 + 16 - (16 >>N)` 

Then the adc value is always increasing where it is supposed to remain the same, right?
Unless 2^N is larger than the value itself.

Reply:
I see now what you are missing:  adc = adc << filter_power; //this is the first measurement
the previous value is not a raw value but is a sum of the window average.
In the end actually that is not even matter other for the fact that will take a while to get to the real value.

the final value is the current adc >> filter_power 
see: 
if (P014_data->chip_id == CHIP_ID_SI7013){
          UserVar[event->BaseVarIndex + 2] = (P014_data->adc) >> P014_FILTER_POWER;
        }

so you can think that the adc contains the sum all of the values in the moving window. 
We remove the average and we add the read value. thus we only add the delta between the average and the current value.

In your example that would be zero. because the current adc value would not be 16 but 16<<N

Reply:
OK, that part is clear now, I was missing the `adc<<=N ` part...
But I do need to do some calculations on what you're actually doing as I don't think this is a moving averaging filter.
A moving averaging filter needs the last N numbers.
This looks more like an IIR filter.